File: Assemblies/ProteanCMS/Tools/Object.vb
Imports System.Configuration
Imports System.Web.Configuration

''' <summary>
''' Protean.Cms.TypeExtensions extends Protean.Tools.TypeExtensions
''' allows handling of various scenarios that return a type.
''' </summary>
''' <remarks></remarks>
Partial Public Class Web
    Public Class TypeExtensions
        Inherits Protean.Tools.TypeExtensions



        Shared Function TypeFromProviderName(ByVal typeFromSpecifiedAssembly As String, ByVal providerName As String, ByVal providerSection As String) As System.Type

            Dim providerConfig As Protean.ProvidersectionHandler = WebConfigurationManager.GetWebApplicationSection(providerSection)
            If providerConfig IsNot Nothing Then
                Dim provider As ProviderSettings = providerConfig.Providers(providerName)
                If provider IsNot Nothing Then
                    Return Protean.Tools.TypeExtensions.Type(typeFromSpecifiedAssembly, provider.Type)
                Else
                    Return Nothing
                End If
            Else
                Return Nothing
            End If

        End Function

        Overloads Shared Function Type(ByVal typeName As String, ByVal assemblyName As String, ByVal assemblyType As String, ByVal providerName As String, ByVal providerSection As String) As System.Type

            If Not String.IsNullOrEmpty(providerName) And Not String.IsNullOrEmpty(providerSection) Then
                Return TypeFromProviderName(typeName, providerName, providerSection)
            ElseIf Not String.IsNullOrEmpty(assemblyName) Then
                Return Protean.Tools.TypeExtensions.Type(typeName, assemblyName, assemblyType)
            Else
                ' Need to avoid going back to Protean.Tools, so just repliacte here
                Return System.Type.GetType(typeName, True, True)
            End If

        End Function

    End Class
End Class

File: Assemblies/Protean.CMS/tools/Object.cs
using System;
using System.Web.Configuration;

namespace Protean
{

    /// <summary>
/// Protean.Cms.TypeExtensions extends Protean.Tools.TypeExtensions
/// allows handling of various scenarios that return a type.
/// </summary>
/// <remarks></remarks>
    public class Web
    {
        public class TypeExtensions : Tools.TypeExtensions
        {

            public static Type TypeFromProviderName(string typeFromSpecifiedAssembly, string providerName, string providerSection)
            {

                Protean.ProviderSectionHandler providerConfig = (Protean.ProviderSectionHandler)WebConfigurationManager.GetWebApplicationSection(providerSection);
                if (providerConfig is not null)
                {
                    var provider = providerConfig.Providers[providerName];
                    if (provider is not null)
                    {
                        return Tools.TypeExtensions.Type(typeFromSpecifiedAssembly, provider.Type);
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }

            }

            public static Type Type(string typeName, string assemblyName, string assemblyType, string providerName, string providerSection)
            {

                if (!string.IsNullOrEmpty(providerName) & !string.IsNullOrEmpty(providerSection))
                {
                    return TypeFromProviderName(typeName, providerName, providerSection);
                }
                else if (!string.IsNullOrEmpty(assemblyName))
                {
                    return Type(typeName, assemblyName, assemblyType);
                }
                else
                {
                    // Need to avoid going back to Protean.Tools, so just repliacte here
                    return System.Type.GetType(typeName, true, true);
                }

            }

        }
    }
}